### DISCUSSION TITLE
How to access server file from VsCode WEB extension for WEB VsCode hosted on my server ?

### DISCUSSION THREAD

**Discussion by AlexDzhus1994** [microsoft/vscode-discussions#839] [Extension Development QnA] [answered]
Good afternoon, our team faced with an issue during VsCode Web Extension development.
We plan to provide a Web version of VsCode hosted on our server to clients.
The user works in the same server (and file system) that hosts the VsCode.

We are trying to create an extension that would read temp credentials from a file filled by a backend app.
Any other way of communication between the extension and the backend app except of public http request works for us.
Public http request doesn't work as it requires credentials hardcoding in the extension to access the server.
We know that VsCode web extensions are running in the isolated browser sandbox, but still some extensions have the access to backend files.
As an example "AWS Toolkit" extension reads AWS credentials from "~/.aws/credentials" backend file with no issues.

We tried a couple of ways to access backend resources from the web extension (shown below), but none of them works.
Can you please advise how can we reach our goal ?

Result of the code execution:
**tryEnvVars** - return zero as env variables count, when "env" command from VsCode CLI returns a lot of env variables.
**tryFileExists** - returns "FILE NOT FOUND"
**tryOpenTextDocument** - doesn't open the document in the editor (it won't solve our problem, but still even this feature doesn't work)
**tryFileCreate** - doesn't create a file
**tryReadDirectory** - doesn't find anything

The code:
```
import * as vscode from 'vscode';

function sleep(ms: number) {
    return new Promise((resolve) => {
      setTimeout(resolve, ms);
    });
}

function tryEnvVars(){
    let envVarsCount = Object.keys(process.env).length;
    vscode.window.showInformationMessage("ENV VARS COUNT: " + envVarsCount.toString());
}

async function tryFileExists(uri: vscode.Uri)
{
    vscode.workspace.fs.stat(uri).then(
        () => vscode.window.showInformationMessage("FILE FOUND, " + uri.toString()),
        err => vscode.window.showInformationMessage("FILE NOT FOUND, " + err.toString())
    );
}

function tryOpenTextDocument(uri: vscode.Uri){
    vscode.workspace.openTextDocument(uri);
}

function tryFileCreate(uri: vscode.Uri){
        let wsedit = new vscode.WorkspaceEdit();
        wsedit.createFile(uri, {ignoreIfExists: true});
        vscode.workspace.applyEdit(wsedit);
}

function tryReadDirectory(uri: vscode.Uri){
        vscode.workspace.fs.readDirectory(uri).then(doc => {
            vscode.window.showInformationMessage("FOUND SOMETHIG IN " + uri.toString());
        });
}

export function activate(context: vscode.ExtensionContext) {

    let disposable = vscode.commands.registerCommand('testExtension.helloWorld', async () => {
        
        vscode.window.showInformationMessage("STARTED");

        tryEnvVars();
        await sleep(10000);

        tryFileExists(vscode.Uri.file("/home/coder/TEST_FILE.txt"))
        tryFileExists(vscode.Uri.file("home/coder/TEST_FILE.txt"))
        await sleep(10000);

        tryFileExists(vscode.Uri.parse('file:~/TEST_FILE.txt'))
        tryFileExists(vscode.Uri.parse("file:./TEST_FILE_LOCAL.txt"))
        await sleep(10000);

        tryOpenTextDocument(vscode.Uri.file("/home/coder/TEST_FILE.txt"));
        tryOpenTextDocument(vscode.Uri.file("home/coder/TEST_FILE.txt"));
        await sleep(10000);

        tryFileCreate(vscode.Uri.file("/home/coder/CREATED_FILE_1.txt"));
        tryFileCreate(vscode.Uri.file("home/coder/CREATED_FILE_2.txt"));
        await sleep(10000);

        tryReadDirectory(vscode.Uri.file("/home/coder/"));
        tryReadDirectory(vscode.Uri.file("home/coder/"));
        tryReadDirectory(vscode.Uri.file("home/coder"));
        await sleep(10000);

        tryReadDirectory(vscode.Uri.file("/home/coder/proj/"));
        tryReadDirectory(vscode.Uri.file("home/coder/proj/"));
        tryReadDirectory(vscode.Uri.file("home/coder/proj"));
        await sleep(10000);

        tryReadDirectory(vscode.Uri.file("."));
        tryReadDirectory(vscode.Uri.file("./"));
        tryReadDirectory(vscode.Uri.file("~/"));
        await sleep(10000);

        vscode.window.showInformationMessage("DONE");
    });

    context.subscriptions.push(disposable);
}

export function deactivate() {}
```

**Comment by Ninglo** [comment]
First, we need to confirm whether we have the same understanding of the term `VSCode Web`. In my understanding, `VSCode Web` refers to Web App applications that purely run in a browser environment, such as `vscode.dev` or `github.dev`, while products with a backend environment are `code-server` (similar to products like github codespace). For the former (VSCode Web), we obviously cannot read files on the real machine (because it doesn't even have a backend service!). For the latter (code-server), it supports running regular VSCode plugins (not just VSCode Web plugins), so you just need to develop plugins normally and use the `fs` api provided by nodejs to access files. 
Perhaps these reference links will be useful to you:
* https://code.visualstudio.com/blogs/2022/07/07/vscode-server
* https://code.visualstudio.com/blogs/2021/10/20/vscode-dev
* https://code.visualstudio.com/api/extension-guides/web-extensions

**Reply by AlexDzhus1994** [reply]
@Ninglo , thank you for the reply !
When I go to **Help** -> **About** in the VsCode, it shows me
**code-server: v4.4.0
Code:1.66.2**
So I believe I do use the code server.

I made an experiment, I created 2 extension via the extension generator (sudo yo code), 1st  is Web Extension, 2nd is a regular extension.
Then I created VSIX file from both and installed them to my VsCode (code server I guess)
1st works fine, 2nd was installed, but gives error "Command XXX not found" when I try to trigger it, at the same time it works fine in regular VsCode that is running on my Linux computer.

Can you please advise if any additional steps are required to make none-web extension run on VsCode server ?

**Reply by Ninglo** [reply]
`Command XXX not found`, What commands are specifically mentioned in the error message? It should be noted that the latest version of VSCode is currently 1.82, but your code-server VSCode version is only 1.66. Excessive version difference may be the cause of the problem (more information is needed to prove my guess). One way to eliminate it is to control the scope of the problem, such as creating a new plug-in whose function is only to call the nodejs fs api to read a file and display it through the vscode api, or you can design other solutions based on your understanding. Hope this helps you.

**Reply by Ninglo** [reply]
Generally, there is no difference between running plugins on the code server and local VSCode.

**Reply by Ninglo** [reply]
`https://code.visualstudio.com/api/advanced-topics/remote-extensions#architecture-and-extension-kinds` This document also introduces some notes on code server plug-in development.

**Reply by AlexDzhus1994** [reply]
"What commands are specifically mentioned in the error message".
It shows the name of the command that I declarer in my extension as a main command, basically a command that I call when I press CTRL+SHIFT+P  and select what to run. Currently it is just "Hello World".


"is only to call the nodejs fs api to read a file and display it through the vscode api,"
Currently both extensions from my previous message have nothing except a "Toast" message.
I will provide the code itself in 20 min just in case.

**Reply by Ninglo** [reply, marked as answer]
Using code similar to the following should be able to confirm whether your project can read files on the machine.
```TS
import * as fs from 'fs/promises'
import * as vscode from 'vscode'

const YOUR_FILE_PATH = ''

export async function activate(context) {
    const content = await fs.readFile(YOUR_FILE_PATH, { encoding: 'utf-8' })
    vscode.window.showInformationMessage(`file content is ${content}`)
}
```

**Reply by AlexDzhus1994** [reply]
@Ninglo , it works, thank you so much !
Not sure what was wrong with my original none-web extension, but I re-created it from a scratch and used the code that you provided.
It work as a charm, thank you !
